src/upnp.cpp

```diff
@@ -965,8 +965,31 @@ void upnp::on_upnp_xml(error_code const& e
 		return;
 	}
 
+	// determine IGD version
+	std::string igd_str(p.get_body().begin, p.get_body().end);
+	size_t igd_ver_idx = igd_str.find("urn:schemas-upnp-org:device:InternetGatewayDevice:");
+	if (igd_ver_idx == std::string::npos)
+	{
+		char msg[500];
+		snprintf(msg, sizeof(msg), "missing IGD version information from %s"
+			, d.url.c_str());
+		//log(msg, l);
+		return;
+	}
+	int igd_ver = igd_str[igd_ver_idx + 50] - '0';
+	if (igd_ver < 1 || igd_ver > 2)
+	{
+		char msg[500];
+		snprintf(msg, sizeof(msg), "invaild IGD version from %s, falling back to version 1"
+			, d.url.c_str());
+		log(msg, l);
+		igd_ver = 1;
+	}
+	d.igd_version = igd_ver;
+
 	parse_state s;
-	s.reset("urn:schemas-upnp-org:service:WANIPConnection:1");
+	static std::string wanip_service = std::string("urn:schemas-upnp-org:service:WANIPConnection:") + to_string(d.igd_version).elems;
```

Comment:
I think it would make more sense find_control_url to recognize the string prefix and parse out the version itself. It may even be a good opportunity to merge looking for WANIPConnection and PPPConnection as well.

Currently, there are just two lookups, if the WANIPConnection can't be found, we search for WANPPPConnection instead. See: https://github.com/arvidn/libtorrent/pull/23/files#diff-54a4016ca61d648addae2913638804d9R1004

Even if those two paths aren't merged, doesn't PPPConnection have to support version 2 as well?